Spec/RealtimeClient.swift

```diff
@@ -352,6 +352,478 @@ class RealtimeClient: QuickSpec {
+                    client.setTransportClass(TestProxyTransport.self)
+
+                    waitUntil(timeout: testTimeout) { done in
+                        client.connection.once(.Connected) { stateChange in
+                            expect(stateChange?.reason).to(beNil())
+                            done()
+                        }
+                        client.connect()
+                    }
+
+                    waitUntil(timeout: testTimeout) { done in
+                        let partialDone = AblyTests.splitDone(3, done: done)
+
+                        client.connection.once(.Connecting) { stateChange in
+                            guard let stateChange = stateChange else {
+                                fail("ConnectionStateChange is nil"); partialDone(); return
+                            }
+                            expect(stateChange.previous).to(equal(ARTRealtimeConnectionState.Closed))
+                            expect(stateChange.reason).to(beNil())
+                            partialDone()
+                        }
+
+                        client.connection.once(.Connected) { stateChange in
+                            guard let stateChange = stateChange else {
+                                fail("ConnectionStateChange is nil"); partialDone(); return
+                            }
+                            expect(stateChange.previous).to(equal(ARTRealtimeConnectionState.Connecting))
+                            expect(stateChange.reason).to(beNil())
+                            partialDone()
+                        }
+
+                        client.auth.authorize(nil, options: nil) { tokenDetails, error in
+                            expect(error).to(beNil())
+                            guard let tokenDetails = tokenDetails else {
+                                fail("TokenDetails is nil"); partialDone(); return
+                            }
+
+                            expect(client.connection.state).to(equal(ARTRealtimeConnectionState.Connected))
+                            expect(tokenDetails.token).toNot(equal(testToken))
+
+                            guard let transport = client.transport as? TestProxyTransport else {
+                                fail("TestProxyTransport is not set"); partialDone(); return
+                            }
+                            expect(transport.protocolMessagesSent.filter({ $0.action == .Auth })).to(haveCount(0))
+                            expect(transport.protocolMessagesReceived.filter({ $0.action == .Connected })).to(haveCount(1)) //New transport
+                            partialDone()
+                        }
+
+                        guard let transport = client.transport as? TestProxyTransport else {
+                            fail("First transport is not TestProxyTransport"); done(); return
+                        }
+                        transport.simulateIncomingNormalClose()
```

Comment:
Surely the `authorize` request should only be called once you have confirmed it is indeed in the `CLOSED` state as according to the spec "when the connection is in the DISCONNECTED, SUSPENDED, FAILED, or CLOSED state when auth#authorize...".